### ISSUE TITLE
Support backward hook optimizers in FSDP

### ISSUE THREAD

**Issue by robieta** [pytorch/pytorch#98419]
### 🚀 The feature, motivation and pitch

I'm currently optimizing the [Lightning reference implementation of LLaMA](https://github.com/Lightning-AI/lit-llama) (7B), although the following will be generally applicable to any LLM with high memory pressure. The default configuration (24GB model sharded across 4x40GB A100s) is just on the cusp of being able to run (Between weights, 2 AdamW states, and gradients for the shard **logical** GPU memory caps out around 27GB, although I don't think that captures comms buffers) and the profile shows clear signs of allocator thrashing.

After a bit of hacking I came up with this monstrosity:
```
def add_optimizer_hooks(
    model,
    optimizers: Dict[torch.nn.Parameter, torch.optim.Optimizer],  # Per-parameter optimizers
):
    """Ugly FSDP analog to torch.distributed.optim._apply_optimizer_in_backward
    
    FSDP changes acc_grad every step, so we need to apply this before *each* `backward()`
    call, unlike the normal recipe where we only apply it once.
    """

    param_handles = torch.distributed.fsdp._traversal_utils._get_fsdp_handles(model)
    assert set(model.parameters()) == {i.flat_param for i in param_handles} == set(optimizers.keys())

    # We need to use the post backward stream so updates apply gradients are accumulated
    stream = torch.distributed.fsdp._common_utils._get_module_fsdp_state(model)._streams["post_backward"]

    for h in param_handles:
        # We're going to call this early, so if we don't override to a no-op FSDP proper will call it again and assert fail.
        h.prepare_gradient_for_optim = lambda: None

        p = h.flat_param
        assert hasattr(p, "_post_backward_hook_state")
        fsdp_acc_grad, _ = p._post_backward_hook_state

        def _opt_hook(optimizer, p, h, *_unused):
            assert p._post_backward_called

            with torch.cuda.stream(stream):
                # Use the class to get at `prepare_gradient_for_optim`
                h.__class__.prepare_gradient_for_optim(h)
                assert p.grad is not None
                optimizer.step()
                optimizer.zero_grad(set_to_none=True)  # Cool that this is now the default
                assert p.grad is None

        fsdp_acc_grad.register_hook(functools.partial(_opt_hook, optimizers[p], p, h))
```

<img width="350" alt="Screenshot 2023-04-05 at 8 43 23 AM" src="https://user-images.githubusercontent.com/13089297/230133300-0865ec63-45e5-416c-aa66-091c16c3ef3e.png"> <img width="339" alt="Screenshot 2023-04-05 at 8 43 47 AM" src="https://user-images.githubusercontent.com/13089297/230133388-ab6303bb-a2fa-4c22-baff-a1f075bf5a68.png">

More importantly, that's enough to get out of the high contention regime and decreases the step time close to an order of magnitude. But given how much FSDP internal state I had to crack open to get things running (and I'm sure I missed plenty...) it's really only suitable as a PoC.

CC @rohan-varma @albanD @zdevito

### Alternatives

I know there's been more general discussion of creating optimizers on the fly so there might be a better alternative to the big list of single Tensor optimizers.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by albanD** [comment]
From checking, I see there are already 3 implementations of this in torch.distributed:
- torch.distributed.algorithms.ddp_comm_hooks.
- torch.distributed.algorithms._optimizer_overlap
- torch.distributed.optim.apply_optimizer_in_backward

Maybe one of these work with FSDP?

**Comment by awgu** [comment]
I think @rohan-varma had plans for adding this support in FSDP. Maybe we just need to re-sync on the timeline.

**Comment by rohan-varma** [comment]
Hi @robieta @albanD, sorry for the late follow up here! 

https://github.com/pytorch/pytorch/pull/98667 prototypes the FSDP overlapped optimizer by making it work with `apply_optimizer_in_backward` API. 

It uses many similar ideas as in @robieta's prototype:
- Runs optimizer step in post backward
- set grad to None after optimizer step call

However, had a q on your implementation:
- Since we use `register_hook` API and FSDP uses this API as well separately to do gradient communication in backward, how do we ensure these are appropriately ordered?

Overall, we'd like FSDP optimizer overlap to support the following:
- Usable through apply_optimizer_in_backward API
- Setting grad to None after step() provides the expected memory saving
- Ideally, in CPU offload case, we can run optimizer step on the GPU before offloading params and grads, but this needs some more design discussion. 

@albanD, agreed the current situation with hooks is confusing. Here is a small clarification:
- ddp_comm_hooks: are supposed to be general hooks for gradient communication (i.e. bf16 allreduce)
- _optimizer_overlap: was our first prototype of DDP optimizer overlap, but not the API we are planning to publicly release. We should kill this and remove the code
- apply_optimizer_in_backward is our new optimizer overlap API, which can work for general use cases but is being implemented for DDP / FSDP

**Comment by robieta** [comment]
> Since we use `register_hook` API and FSDP uses this API as well separately to do gradient communication in backward, how do we ensure these are appropriately ordered?

On the host side I rely on the existence of `_post_backward_hook_state` to know that FSDP has gotten a chance to register its hooks. (It *seems* like backward hooks respect registration order...) And of course I had to use the comm stream to make sure device kernels were properly ordered.

I'm very happy that you're thinking of how to pair this with CPU offload. In this case I don't know that you can get much out of parameter offloading since they not only need to be present in the forward, but also the backward due to activation checkpointing. OTOH if you kept the optimizer state on host, sent grads to the host, and sent parameter updates back to device you could save 2 bytes of device memory for every byte round-trip'd. (My experiments indicate that you don't have enough PCI bandwidth to do this for all state, but even offloading a fraction would do wonders.) But yeah, very complex and interesting space so some design work is needed.

**Comment by Xynonners** [comment]
Hi, it seems as though this discussion has stopped for some reason. I'm also interested in how this can be used with an optimizer hook (as I want to do gradient clipping of some sort).

**Comment by mori360** [comment]
Backward hook for optimizers is available at https://github.com/pytorch/pytorch/pull/107063
If you need further support with FSDP2, please open a new issue with updated information.
